File: frontend/src/lib/renewals.ts
/**
 * Typed API client for Phase 4 renewals endpoints.
 *
 * Sibling to lib/claims.ts. Money values (annual_premium,
 * projected_loss_adjustment, prior_annual_premium, loss_adjustment) come back
 * as STRINGS — parse only at render-time.
 *
 * Error shapes mirror claims: 4xx returns { detail } where detail is either a
 * string or a structured object { error, message }.
 */
import { authHeaders } from "@/lib/authFetch";
import { PlacementApiError } from "@/lib/placement";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://127.0.0.1:8000";

// ─── Wire types (match backend/app/api/v1/renewals.py response shapes) ───

export interface RenewalDue {
  policy_id: string;
  policy_number: string | null;
  venue_id: string;
  expiration_date: string;              // ISO date
  annual_premium: string;              // Decimal as string
  loss_ratio: string;                  // Decimal as string
  claim_count: number;
  projected_loss_adjustment: string;   // Decimal as string
}

export interface RenewResult {
  submission: {
    id: string;
    venue_id: string;
    status: string;
    prior_policy_id: string;
    coverage_lines: string[];
    requested_limits: Record<string, unknown>;
    effective_date: string;            // ISO date
  };
  yoy_context: {
    prior_policy_id: string;
    prior_annual_premium: string;      // Decimal as string
    prior_coverage_lines: string[];
    loss_ratio: string;                // Decimal as string
    claim_count: number;
    loss_adjustment: string;           // Decimal as string
  };
}

// ─── Fetch wrapper (mirrors claims.ts call<T> exactly) ───────────────────

async function call<T>(path: string, init: RequestInit = {}): Promise<T> {
  const res = await fetch(`${API_URL}${path}`, {
    ...init,
    headers: {
      "Content-Type": "application/json",
      ...authHeaders(),
      ...(init.headers ?? {}),
    },
  });
  if (!res.ok) {
    let detail: unknown = null;
    try {
      const body = await res.json();
      detail = body?.detail ?? body;
    } catch {
      detail = await res.text().catch(() => "");
    }
    const msg = typeof detail === "string"
      ? detail
      : (detail as { message?: string })?.message ?? `HTTP ${res.status}`;
    throw new PlacementApiError(res.status, msg, detail);
  }
  return (await res.json()) as T;
}

// ─── Endpoints ──────────────────────────────────────────────────────────

export const renewalsApi = {
  /** List policies expiring within `withinDays` days (default 60). */
  due: (withinDays = 60): Promise<RenewalDue[]> =>
    call<RenewalDue[]>(`/api/renewals/due?within_days=${withinDays}`),

  /** Create a renewal submission for a policy. */
  renew: (policyId: string, effectiveDate: string): Promise<RenewResult> =>
    call<RenewResult>(`/api/policies/${policyId}/renew`, {
      method: "POST",
      body: JSON.stringify({ effective_date: effectiveDate }),
    }),
};

File: backend/app/api/v1/renewals.py
"""FastAPI endpoints for Phase 4 renewals. Mounted at /api by main.py.
Broker-gated. Error mapping mirrors the claims/policies routers:
  RenewalsError -> 400, InvalidTransitionError -> 422."""
from __future__ import annotations

from decimal import Decimal

from datetime import date, timedelta
from typing import NoReturn

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlmodel import Session, select

from app.api.v1.placement import _broker_user_id
from app.auth import require_broker
from app.database import get_session
from app.lifecycles import InvalidTransitionError
from app.models import Policy
from app.money import usd_to_json
from app.services.renewals import (
    RenewalsError,
    build_experience_years_for_policy,
    create_renewal,
    find_live_renewal,
)
from app.underwriting.experience_rating import compute_experience_mod

router = APIRouter()


def _map_service_error(e: Exception) -> NoReturn:
    if isinstance(e, InvalidTransitionError):
        raise HTTPException(
            status_code=422,
            detail={"error": "invalid_transition", "message": str(e)},
        )
    if isinstance(e, RenewalsError):
        raise HTTPException(
            status_code=400,
            detail={"error": "renewals_error", "message": str(e)},
        )
    raise e


class RenewBody(BaseModel):
    effective_date: date


@router.get("/renewals/due", dependencies=[Depends(require_broker)])
def renewals_due(
    within_days: int = 60,
    session: Session = Depends(get_session),
) -> list[dict]:
    cutoff = date.today() + timedelta(days=within_days)
    rows = session.exec(
        select(Policy)
        .where(Policy.status == "active")
        .where(Policy.expiration_date <= cutoff)
        .order_by(Policy.expiration_date)
    )
    out: list[dict] = []
    for pol in rows:
        # A policy with a renewal already in flight (or already bound) is no
        # longer "due" — it's being worked. Surfacing it would nag the broker
        # forever and invite a duplicate renewal.
        if find_live_renewal(session, pol.id) is not None:
            continue
        years = build_experience_years_for_policy(
            session, pol.id,
            annual_premium=pol.annual_premium,
            years_back=0,
        )
        mod = compute_experience_mod(years)
        out.append({
            "policy_id": pol.id,
            "policy_number": pol.policy_number,
            "venue_id": pol.venue_id,
            "expiration_date": pol.expiration_date.isoformat(),
            "annual_premium": usd_to_json(pol.annual_premium),
            "loss_ratio": str(mod.experience_lr),
            "claim_count": mod.claim_count,
            "credibility_z": str(mod.credibility_z.quantize(Decimal("0.0001"))),
            "projected_loss_adjustment": str(mod.mod),
        })
    return out


@router.post(
    "/policies/{policy_id}/renew",
    status_code=201,
    dependencies=[Depends(require_broker)],
)
def renew_policy(
    policy_id: str,
    body: RenewBody,
    session: Session = Depends(get_session),
    actor_id: str = Depends(_broker_user_id),
) -> dict:
    try:
        prior = session.get(Policy, policy_id)
        if prior is None:
            raise HTTPException(status_code=404, detail=f"Policy {policy_id} not found")
        years = build_experience_years_for_policy(
            session, policy_id,
            annual_premium=prior.annual_premium,
            years_back=0,
        )
        mod = compute_experience_mod(years)
        sub = create_renewal(
            session,
            policy_id,
            effective_date=body.effective_date,
            actor_id=actor_id,
        )
        session.commit()
        session.refresh(sub)
    except (RenewalsError, InvalidTransitionError) as e:
        session.rollback()
        _map_service_error(e)

    return {
        "submission": {
            "id": sub.id,
            "venue_id": sub.venue_id,
            "status": sub.status,
            "prior_policy_id": sub.prior_policy_id,
            "coverage_lines": sub.coverage_lines,
            "requested_limits": sub.requested_limits,
            "effective_date": sub.effective_date.isoformat(),
        },
        "yoy_context": {
            "prior_policy_id": policy_id,
            "prior_annual_premium": usd_to_json(prior.annual_premium),
            "prior_coverage_lines": prior.coverage_lines,
            "loss_ratio": str(mod.experience_lr),
            "claim_count": mod.claim_count,
            "credibility_z": str(mod.credibility_z.quantize(Decimal("0.0001"))),
            "loss_adjustment": str(mod.mod),
        },
    }
